compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -1933,7 +1933,19 @@ impl<'a> Parser<'a> {
         }
         match self.parse_expr_res(Restrictions::CONST_EXPR, None) {
             Ok(expr) => {
-                if token::Comma == self.token.kind || self.token.kind.should_end_const_arg() {
+                // Find a mistake like `MyTrait<Assoc == S::Assoc>`.
+                if token::EqEq == snapshot.token.kind {
```

Comment:
I feel like this is the wrong place for this. Haven't dug into this, but I would have expected this to be in the same place as the associated type parsing code? There, we just "accept" with an error `==` when we really want `=`.

I guess this is potentially "ambiguous" if `Assoc == S::Assoc` gets treated as a const arg, which I guess is what currently happens.

Is there anything we can to be smarter here? Would we theoretically be allowed to have a variable named `Assoc`? I'm trying to think of if we had complete knowledge of all traits, types, variables, etc. in scope, are there any cases where we can know if this should be a bracketed const expr arg or an associated type binding?

Reply:
I got somewhat confused as to why we didn't have a `==` binop here, but looking at the surrounding code I can now see that we discard whatever op this was and what is being parsed here is the RHS, namely a single path. So I'm getting around to thinking that this is not a bad place to make this check.